### PR TITLE
bpo-45455: Set `func.__doc__` lazily.

### PR DESCRIPTION
**Pull request by methane**
https://bugs.python.org/issue45455

### CODE REVIEW

File: Objects/funcobject.c [hunk 1]
```diff
@@ -510,6 +494,39 @@ func_set_annotations(PyFunctionObject *op, PyObject *value, void *Py_UNUSED(igno
...
+static PyObject *
+func_get_doc(PyFunctionObject *op, void *Py_UNUSED(ignored))
+{
+    if (op->func_doc == NULL) {
+        PyObject *consts = ((PyCodeObject*)(op->func_code))->co_consts;
+        assert(PyTuple_Check(consts));
+        PyObject *doc;
+
+        if (PyTuple_Size(consts) >= 1) {
+            doc = PyTuple_GetItem(consts, 0);
```
Comment (serhiy-storchaka) [inline Objects/funcobject.c]: You can use macros: PyTuple_GET_SIZE, PyTuple_GET_ITEM.

**Review by serhiy-storchaka — APPROVED** [review]
Yesterday I experimented with a patch which added the `co_doc` property and `_PyCode_GetDoc()` for extracting a docstring from code's constants. I was not sure in the usefulness of that feature, seems there are no use of `co_doc` in the Python code, and there is exactly one use of `_PyCode_GetDoc()` in the C code.

But your PR looks more interesting.

**Comment by warsaw** [comment]
Do you have a bug #?  Is the motivation to reduce memory overhead and performance with the docstring isn't used?

**Comment by methane** [comment]
@warsaw This PR is spin-off idea of bpo-36521. Motivation is faster function creation (including lambda) when docstring is not used.

But performance gain is only 3~5% in micro benchmark and having several backward incompatibility.

**Comment by rhettinger** [comment]
I don't see the benefit.  We still use the same amount of the memory.  The pointer just gets copied later rather than sooner. AFAICT, at best it saves a single INCREF.

**Comment by methane** [comment]
Yes. This have only 3~5% performance gain on microbenchmark.

**Pull request closed without merging**